epios/post_process.py

```diff
@@ -0,0 +1,1423 @@
+            plt.xlim(0, max(time_sample))
+            plt.ylim(0, len(self.demo_data))
+            plt.title('Number of infection in the sample')
+            if saving_path_sampling:
+                plt.savefig(saving_path_sampling)
+        res = []
+        res.append(time_sample)
+        res.append(infected_rate)
+        # Output the results for comparison use
+        self.result = infected_rate
+        return res
+
+    def sampled_non_responder(self, sampling_method, sample_size, time_sample, nonresprate,
+                              gen_plot: bool = False, saving_path_sampling=None, sampling_percentage=0.1,
+                              proportion=0.01, threshold=None, num_age_group=17, age_group_width=5,
+                              data_store_path='./input/'):
+        '''
+        This is a method to generate the sampled result and plot a figure when considering non-responders
+        This method should not be used directly, it is integrated within the __callable__ method
+
+        '''
+        if sampling_method == 'AgeRegion':  # Both age and region stratification
+
+            # Only Random sample strategies
+            infected_rate = []
+            for i in range(len(time_sample)):  # Sample again at each time point
+                if i == 0:
+                    sampler_class = SamplerAgeRegion(data=self.demo_data, data_store_path=data_store_path,
+                                                     num_age_group=num_age_group, age_group_width=age_group_width)
+                else:
+                    sampler_class = SamplerAgeRegion(data_store_path=data_store_path, pre_process=False,
+                                                     num_age_group=num_age_group, age_group_width=age_group_width)
+                try:
+                    people = sampler_class.sample(sample_size=sample_size, additional_sample=additional_sample)
+                except NameError:
+                    people = sampler_class.sample(sample_size=sample_size)
+
+                # Get the results of people sampled
+                X = SamplingMaker(nonresprate=nonresprate, keeptrack=True, TheData=self.time_data,
+                                  false_positive=0, false_negative=0, threshold=None)
+                ite = X([time_sample[i]], people)
+
+                # After each sample, now deal with the additional samples
+                try:
+                    # For the first time sampled, there is no variable additional sample defined,
+                    # therefore, NameError would be raised
+                    # Then it will go to the 'except NameError' line
+                    additional_sample = np.array(additional_sample)
+                    if additional_sample.sum() == 0:
+                        # When there is no additional samples needed after one sample,
+                        # we will also skip the following bit directly output the infection rate
+                        # by raising the NameError
```

Comment:
The reason I did this is to improve the efficiency of the code. Without any non-responders, we can skip the following bit directly to calculate the infected rate. When we have non-responders, we need to run the following bit to calculate a precise infected rate. Reason: We have additionally sampled people from some specific age-regional group since many people from this group do not respond. If we still want to keep the age and region stratified, we have to do the following parts to eliminate the bias introduced by these additional samples.